### PR TITLE
Nipopow bootstrapping from multiple peers

### PR DESCRIPTION
**Pull request by kushti**
In this PR, NiPoPoW proofs first downloaded from multiple peers and compared, then with the best on headers-chain is initialized

### CODE REVIEW

File: src/main/scala/org/ergoplatform/nodeView/history/ErgoHistory.scala [hunk 1]
```diff
@@ -314,6 +314,23 @@ object ErgoHistory extends ScorexLogging {
 
     repairIfNeeded(history)
 
+    // temporary hack which is injecting nipopow proof to the database to make it possible to bootstrap with
+    // nipopows + utxo set snapshot soon after 5.0.13 release
+    // todo: remove after height 1,096,693 on the mainnet
+    val bestHeaderHeight = history.headersHeight
+    if (bestHeaderHeight > 1054000 && bestHeaderHeight < 1096693) { //history.readPopowProofBytesFromDb().isDefined){
```
Comment (aslesarenko) [inline src/main/scala/org/ergoplatform/nodeView/history/ErgoHistory.scala:321]: commented code

File: src/main/scala/org/ergoplatform/nodeView/history/ErgoHistory.scala [hunk 2]
```diff
@@ -314,6 +314,23 @@ object ErgoHistory extends ScorexLogging {
...
+      history.popowProofBytes(
+        history.P2PNipopowProofM,
+        history.P2PNipopowProofK,
+        Some(ModifierId @@ "25a11667e38e62412522c062d90b073afd9ed9551080ff4e0a67d1757ce18b98")) match {
```
Comment (aslesarenko) [inline src/main/scala/org/ergoplatform/nodeView/history/ErgoHistory.scala:325]: Please create a named constant with a description where this id come from.

**Review by aslesarenko — CHANGES_REQUESTED** [review]

File: src/main/scala/org/ergoplatform/nodeView/history/storage/modifierprocessors/HeadersProcessor.scala [hunk 3]
```diff
@@ -165,13 +165,14 @@ trait HeadersProcessor extends ToDownloadProcessor with PopowProcessor with Scor
...
+    val timeToTakeNipopowProof: Boolean = if (settings.nodeSettings.nipopowSettings.nipopowBootstrap) {
       // currently, the node is not taking NiPoPoW proof if was bootstrapped via a NiPoPoW itself,
       // as no extension sections (with interlinks) available for headers downloaded via nipopows
       false
     } else  {
-     this.isHeadersChainSynced &&
-        h.height % MakeSnapshotEvery == MakeSnapshotEvery - 1 - Constants.LastHeadersInContext
+      val makeSnapshotEvery = chainSettings.makeSnapshotEvery
+      this.isHeadersChainSynced &&
+        h.height % makeSnapshotEvery == makeSnapshotEvery - 1 - Constants.LastHeadersInContext
```
[the remark this answers is not in the document]
Comment (kushti) [reply src/main/scala/org/ergoplatform/nodeView/history/storage/modifierprocessors/HeadersProcessor.scala:175]: done

File: src/main/scala/org/ergoplatform/nodeView/history/ErgoHistory.scala (hunk 2 above)
Comment (kushti) [reply src/main/scala/org/ergoplatform/nodeView/history/ErgoHistory.scala:325]: done

File: src/main/scala/org/ergoplatform/nodeView/history/ErgoHistory.scala [hunk 4]
```diff
@@ -314,6 +314,23 @@ object ErgoHistory extends ScorexLogging {
...
+        case Success(proofBytes) =>
+          log.info("Writing nipopow proof bytes for height 1,044,469")
```
[the remark this answers is not in the document]
Comment (kushti) [reply src/main/scala/org/ergoplatform/nodeView/history/ErgoHistory.scala:327]: added comment why this height. Named constant could be an overkill, as the code will be removed in ~2 months

**Review by aslesarenko — APPROVED** [review]

**Pull request merged**